File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/MaskCharacterFabric.java
package com.vicmikhailau.maskededittext;


class MaskCharacterFabric {
    private static final char ANYTHING_KEY = '*';
    private static final char DIGIT_KEY = '#';
    private static final char UPPERCASE_KEY = 'U';
    private static final char LOWERCASE_KEY = 'L';
    private static final char ALPHA_NUMERIC_KEY = 'A';
    private static final char CHARACTER_KEY = '?';
    private static final char HEX_KEY = 'H';

    MaskCharacter buildCharacter(char ch) {
        MaskCharacter result = null;
        switch (ch) {
            case ANYTHING_KEY:
                result = new LiteralCharacter();
                break;
            case DIGIT_KEY:
                result = new DigitCharacter();
                break;
            case UPPERCASE_KEY:
                result = new UpperCaseCharacter();
                break;
            case LOWERCASE_KEY:
                result = new LowerCaseCharacter();
                break;
            case ALPHA_NUMERIC_KEY:
                result = new AlphaNumericCharacter();
                break;
            case CHARACTER_KEY:
                result = new LetterCharacter();
                break;
            case HEX_KEY:
                result = new HexCharacter();
                break;
            default: {
                result = new LiteralCharacter(ch);
            }

        }
        return result;
    }
}

File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/MaskCharacterFabric.kt
package com.vicmikhailau.maskededittext


internal class MaskCharacterFabric {

    fun buildCharacter(ch: Char): MaskCharacter? {
        return when (ch) {
            ANYTHING_KEY -> LiteralCharacter()
            DIGIT_KEY -> DigitCharacter()
            UPPERCASE_KEY -> UpperCaseCharacter()
            LOWERCASE_KEY -> LowerCaseCharacter()
            ALPHA_NUMERIC_KEY -> AlphaNumericCharacter()
            CHARACTER_KEY -> LetterCharacter()
            HEX_KEY -> HexCharacter()
            else -> {
                LiteralCharacter(ch)
            }
        }
    }

    companion object {
        private const val ANYTHING_KEY = '*'
        private const val DIGIT_KEY = '#'
        private const val UPPERCASE_KEY = 'U'
        private const val LOWERCASE_KEY = 'L'
        private const val ALPHA_NUMERIC_KEY = 'A'
        private const val CHARACTER_KEY = '?'
        private const val HEX_KEY = 'H'
    }
}

